README.rst
nirum-python-jose
=================

This package experiments with an extended Nirum_ HTTP transport which
authorizes clients using JOSE_ (JSON Object Signing and Encryption).
It proposes a new extension of the existing Nirum HTTP transport and
also provides its minimum implementation in Python.

It purposes to authorizes a client, not a server.  In this scheme, a client
assumes a server is trustworthy; this kind of trust is usually made through
a domain name and HTTPS.

An alternative to this scheme is TLS client authentication.  It's more
standard-ish way (which means more portable) to authorize a client.
As it is a TLS-level solution, the vanilla Nirum HTTP transport is agnostic
to it.  However, if a node dealing with TLS thing does not support TLS client
authentication, it is difficult to adapt.  For example, if your server itself
exposes only an HTTP and an L7 in front of that proxies its upstream HTTP to
the public HTTPS, the L7 should support TLS client as well.

On the other hand, this JOSE-based scheme works well with any kind of L4, L7,
or HTTP(S) proxies.  Signing and verification is done upon an application layer.
On the HTTP(S) level, it is just substituting ``Content-Type`` from
``application/json`` to ``application/jose``.

.. _Nirum: http://nirum.org/
.. _JOSE: https://www.iana.org/assignments/jose/jose.xhtml


Client
------

A client code need to use the ``nirum_jose.client.SigningHttpTransport`` class.
Its constructor is mostly compatible with ``nirum_http.HttpTransport`` except
it also takes a shared secret and an algorithm name to use:

.. code-block:: python

   from nirum_jose.client import SigningHttpTransport
   from your_schema import YourService

   transport = SigningHttpTransport(
       'https://example.com/',
       'secret key',
       algorithm='HS256'
   )
   client = YourService.Client(transport)


Server
------

A server code need to use the ``nirum_jose.server.SigningWsgiApp`` instead of
``nirum_wsgi.WsgiApp``.  Its constructor takes a service instance and a shared
secret, again.  (Note that it does not take ``allowed_origins`` or
``allowed_headers``; Configuring CORS is unsupported.)

.. code-block:: python

   from nirum_jose.server import SigningWsgiApp
   from your_app import YourServiceImpl

   service = YourServiceImpl()
   app = SigningWsgiApp(service, 'secret key')


Limitations
-----------

Since it is still experimental, it lacks features and has several limits when
it's compared to the vanilla HTTP transport:

- Python 3.6 or higher is required.
- Service interfaces generated by older versions of the compiler
  than Nirum 0.4.0 are unsupported.
- A server currently can have only one secret key.  This means all clients
  to a server need to share a single secret key together.  It's also why
  we explain it *authorizes* clients, not *authenticates*.
- REST-style URI routes are unsupported; all ``@http-resource`` annotations
  are ignored.
- CORS configuration is unsupported.
- As it is no more vanilla JSON, it is hard to make a request for debug purpose
  using tools like ``curl``.


CHANGES.rst
Changelog
---------

Version 0.1.0
'''''''''''''

To be released.


nirum_jose/client.py
import io
import json
import logging
from typing import Any, Mapping, Sequence, Tuple, Union
import urllib.request

from jose.jws import sign
from nirum.exc import UnexpectedNirumResponseError
from nirum.transport import Transport

__all__ = 'SigningHttpTransport',


class SigningHttpTransport(Transport):

    logger = logging.getLogger(f'{__name__}.SigningHttpTransport')

    def __init__(self, url: str, secret: str, algorithm: str) -> None:
        self.url = url
        self.secret = secret
        self.algorithm = algorithm

    def call(
        self,
        method_name: str,
        payload: Any,
        service_annotations: Mapping[str, Union[str, int, None]],
        method_annotations: Mapping[str, Union[str, int, None]],
        parameter_annotations: Mapping[
            str,
            Mapping[str, Union[str, int, None]]
        ]
    ) -> Tuple[bool, Any]:
        logger = self.logger.getChild('call')
        signed_payload = sign(
            {'_method': method_name, **payload},
            self.secret,
            algorithm=self.algorithm
        )
        request = urllib.request.Request(
            self.url,
            data=signed_payload,
            headers={
                'Accept': 'application/json',
                'Content-Type': 'application/jose',
            },
        )
        logger.debug(
            'An HTTP request for %s():\n%s %s\n%s\n\n%s',
            method_name, request.get_method(), request.full_url,
            request.headers, request.data
        )
        response = urllib.request.urlopen(request)
        try:
            content = json.load(
                io.TextIOWrapper(response, 'utf-8')  # type: ignore
            )
        except ValueError:
            response.seek(0)
            raise UnexpectedNirumResponseError(response.read().decode())
        status = response.status  # type: ignore
        return 200 <= status < 400, content


nirum_jose/wsgi_typing.py
# Copied from: https://gist.github.com/dahlia/e618b303ab64c0e7e32b1f7957ca283a
import types
from typing import (Any, Callable, Iterable, Mapping,
                    Optional, Sequence, Tuple, Union)

__all__ = ('Environ', 'ExcInfo', 'ResponseStream', 'StartResponse',
           'StartResponseCallable', 'StartResponseCallableWithExcInfo',
           'WriteCallable', 'WsgiApplication')


# https://www.python.org/dev/peps/pep-3333/#the-write-callable
WriteCallable = Callable[[bytes], Any]

# https://docs.python.org/3/library/sys.html#sys.exc_info
ExcInfo = Tuple[type, BaseException, types.TracebackType]

# https://www.python.org/dev/peps/pep-3333/#the-start-response-callable
StartResponseCallable = Callable[
    [
        str,  # status
        Sequence[Tuple[str, str]],  # response headers
    ],
    WriteCallable  # write() callable
]
StartResponseCallableWithExcInfo = Callable[
    [
        str,  # status
        Sequence[Tuple[str, str]],  # response headers
        Optional[ExcInfo]  # exc_info
    ],
    WriteCallable  # write() callable
]
StartResponse = Union[StartResponseCallable, StartResponseCallableWithExcInfo]

# https://www.python.org/dev/peps/pep-3333/#environ-variables
Environ = Mapping[str, object]

# https://www.python.org/dev/peps/pep-3333/#buffering-and-streaming
ResponseStream = Iterable[bytes]

# https://www.python.org/dev/peps/pep-3333/#specification-details
WsgiApplication = Callable[
    [
        Environ,  # environ
        StartResponse,  # start_response() function
    ],
    ResponseStream  # bytestring chunks
]


nirum_jose/server.py
import codecs
import collections.abc
import json
import logging
from typing import BinaryIO, MutableMapping, MutableSet, Sequence, Tuple, cast
from jose.constants import ALGORITHMS
from jose.exceptions import JWSError
from jose.jws import verify
from nirum.service import Service

from .wsgi_typing import Environ, ResponseStream, StartResponseCallable

__all__ = 'SigningWsgiApp',


class SigningWsgiApp:

    logger = logging.getLogger(f'{__name__}.SigningWsgiApp')

    def __init__(self,
                 service: Service,
                 secret: str,
                 algorithms: Sequence[str]=ALGORITHMS.SUPPORTED) -> None:
        if hasattr(type(service), '__nirum_schema_version__'):
            raise TypeError(
                'service interfaces generated by older versions of '
                'the compiler than Nirum 0.4.0 are unsupported'
            )
        self.service = service
        self.secret = secret
        self.algorithms = algorithms

    def error(self,
              code: int,
              message: str,
              environ: Environ,
              start_response: StartResponseCallable) -> ResponseStream:
        statuses = {
            400: 'Bad Request',
            405: 'Method Not Allowed',
        }
        start_response(
            f'{code} {statuses[code]}',
            [('Content-Type', 'text/plain')]
        )
        return codecs.iterencode(message, 'utf-8')

    def get_method_meta(self, facial_name: str, attr: str):
        cls = type(self.service)
        for cls in type(self.service).__mro__:
            if not hasattr(cls, facial_name):
                continue
            method = getattr(cls, facial_name)
            try:
                return getattr(method, attr)
            except AttributeError:
                pass
        assert False, 'could not find the method meta; please report this bug'

    def wsgi_app(self,
                 environ: Environ,
                 start_response: StartResponseCallable) -> ResponseStream:
        if environ.get('REQUEST_METHOD') != 'POST':
            return self.error(405, 'Only POST is supported.',
                              environ, start_response)
        mimetype = environ.get(
            'CONTENT_TYPE',
            environ.get('HTTP_CONTENT_TYPE', 'application/jose')
        )
        if mimetype != 'application/jose':
            return self.error(400, 'Content-Type must be application/jose.',
                              environ, start_response)
        content: BinaryIO = cast(BinaryIO, environ['wsgi.input'])
        signed_payload = content.read()
        try:
            payload_bytes = verify(
                signed_payload,
                self.secret,
                algorithms=self.algorithms
            )
        except JWSError as e:
            self.logger.getChild('wsgi_app').exception('%s', e)
            return self.error(400, f'The signature is failed to verify:\n{e}',
                              environ, start_response)
        payload: MutableMapping[str, object] = \
            json.loads(payload_bytes.decode('utf-8'))
        if not isinstance(payload, collections.abc.Mapping):
            return self.error(400, f'The payload must be a JSON object.',
                              environ, start_response)
        try:
            behind_method_name = payload.pop('_method')
        except KeyError:
            return self.error(400, 'The payload lacks "_method" field.',
                              environ, start_response)
        service_methods = self.service.__nirum_method_names__.behind_names
        try:
            facial_method_name = service_methods[behind_method_name]
        except KeyError:
            return self.error(
                400,
                f'There is no such method: {behind_method_name}().',
                environ, start_response
            )
        cls = type(self.service)
        logger = logging.getLogger(
            f'{cls.__module__}.{cls.__qualname__}.{facial_method_name}'
        )
        deserialize_args = self.get_method_meta(
            facial_method_name,
            '__nirum_deserialize_arguments__'
        )
        errors: MutableSet[Tuple[str, str]] = set()
        args = deserialize_args(payload, lambda ef, em: errors.add((ef, em)))
        if errors:
            return self.error(
                400, ''.join(f'{f}: {m}\n' for (f, m) in errors),
                environ, start_response
            )
        try:
            result = getattr(self.service, facial_method_name)(**args)
        except Exception as e:
            serialize_error = self.get_method_meta(
                facial_method_name,
                '__nirum_serialize_error__'
            )
            if serialize_error is None:
                logger.exception(
                    'The %s() method%s has no error type, but an exception '
                    'has been raised:\n%s',
                    facial_method_name,
                    (
                        ''
                        if behind_method_name == facial_method_name
                        else f' (i.e., {behind_method_name})'
                    ),
                    e
                )
                raise
            try:
                response_payload = serialize_error(e)
            except TypeError as te:
                logger.exception(
                    'The type of a exception value does not match to '
                    'the error type of %s() method%s:\n%s',
                    facial_method_name,
                    (
                        ''
                        if behind_method_name == facial_method_name
                        else f' (i.e., {behind_method_name})'
                    ),
                    te
                )
                raise
            start_response(
                '400 Bad Request',
                [('Content-Type', 'application/json')]
            )
            return codecs.iterencode(json.dumps(response_payload), 'utf-8')
        serialize_result = self.get_method_meta(
            facial_method_name,
            '__nirum_serialize_result__'
        )
        response_payload = None
        if serialize_result is not None:
            try:
                response_payload = serialize_result(result)
            except TypeError as e:
                logger.exception(
                    'The type of a return value does not match to '
                    'the return type of %s() method%s:\n%s',
                    facial_method_name,
                    (
                        ''
                        if behind_method_name == facial_method_name
                        else f' (i.e., {behind_method_name})'
                    ),
                    e
                )
                raise
        elif result is not None:
            logger.error(
                'The %s() method%s has no return type, but a value is '
                'returned; a return value is ignored.',
                facial_method_name,
                (
                    ''
                    if behind_method_name == facial_method_name
                    else f' (i.e., {behind_method_name})'
                )
            )
        start_response('200 OK', [('Content-Type', 'application/json')])
        return codecs.iterencode(json.dumps(response_payload), 'utf-8')

    def __call__(self,
                 environ: Environ,
                 start_response: StartResponseCallable) -> ResponseStream:
        return self.wsgi_app(environ, start_response)
